### PR TITLE
Fix bug with task name as substring of task name.

### LINKED ISSUE #3664: conditional prerequisite bug
**Describe the bug**

Diagnosed after [user bug report on Discourse forum](https://cylc.discourse.group/t/suite-shutting-down-with-name-archive-bool-is-not-defined-error/254/3)

Construction of conditional prerequisite expression fails for `NAME | preNAME => blah` due to overly simplistic string replacement in the `Prerequisite.set_condition()` method of `prerequisite.py`:
```python
# Make a Python expression so we can eval() the logic.
for message in self.satisfied:
    expr = expr.replace(self.MESSAGE_TEMPLATE % message,
                        self.SATISFIED_TEMPLATE % message)
```
**Release version(s) and/or repository branch(es) affected?**

- 7.8.x
- master (Cylc 8)

**Steps to reproduce the bug**

This simple case even fails validation (some only fail at runtime because we don't evaluate all prerequisites during validation):
```ini
[scheduling]
   [[graph]]
      R1 = "foo | xfoo => bar"
```
Validation:
```
$ cylc val foo
name 'xbool' is not defined
SuiteConfigError: bar: failed to evaluate triggers.
```

This probably warrants another Cylc-7 release too :grimacing: 

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by hjoliver**
These changes close #3664

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and  `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

File: cylc/flow/prerequisite.py [hunk 1]
```diff
@@ -115,16 +116,35 @@ def get_raw_conditional_expression(self):
...
         self._all_satisfied = None
         if '|' in expr:
             # Make a Python expression so we can eval() the logic.
             for message in self.satisfied:
-                expr = expr.replace(self.MESSAGE_TEMPLATE % message,
-                                    self.SATISFIED_TEMPLATE % message)
+                # Use '\b' in case one task name is a substring of another
+                # and escape special chars ('.', timezone '+') in task IDs.
+                expr = re.sub(
+                    r"\b" + re.escape(self.MESSAGE_TEMPLATE % message) + r"\b",
```
Comment (kinow) [inline cylc/flow/prerequisite.py:144]: f-string for cylc-8.x branch only (7.8.x probably needs to stay as-is)?

```suggestion
                    fr"\b{re.escape(self.MESSAGE_TEMPLATE % message)}\b",
```

**Review by kinow — APPROVED** [review]
Tested on `master` and got a strange `xfoo`.

```py
['bool(self.satisfied[("foo", "1", "succeeded")])', 'bool(self.satisfied[("xfoo", "1", "succeeded")])']
```

Switched to this branch, and now got the correct result as per doctests.

```py
['bool(self.satisfied[("foo", "1", "succeeded")])', 'bool(self.satisfied[("xfoo", "1", "succeeded")])']
```

Minor comment, but looks good to me! Thanks @hjoliver

File: cylc/flow/prerequisite.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/prerequisite.py:144]: Huh, I wasn't sure if I could nested templating like that!.

**Comment by hjoliver** [comment]
> Tested on master and got a strange xfoo.

I think you pasted the wrong line there? The error case should have `xbool` in it, due to a bad string replacement.

**Comment by kinow** [comment]
> > Tested on master and got a strange xfoo.
> 
> I think you pasted the wrong line there? The error case should have `xbool` in it, due to a bad string replacement.

Ops, copied from the wrong section from the terminal output, fixed! Thanks.

File: cylc/flow/prerequisite.py (hunk 1 above)
Comment (hjoliver) [reply cylc/flow/prerequisite.py:144]: But it does work, so I'll take it.

**Review by dwsutherland — APPROVED** [review]
Reproduced bug on master, couldn't on this branch 👍

**Pull request merged**